dartagnan/src/main/java/com/dat3m/dartagnan/program/analysis/alias/InclusionBasedPointerAnalysis.java

```diff
@@ -435,56 +471,119 @@ private Variable newVariable(String name) {
-        // Cases of 'v == variable' or 'v == edge.source' require recursion.
-        final var stack = new Stack<IncludeEdge>();
-        stack.push(edge);
-        while (!stack.empty()) {
-            final IncludeEdge e = stack.pop();
-            for (final IncludeEdge edgeBefore : List.copyOf(e.source.includes)) {
-                final IncludeEdge joinedEdge = tryInsertIncludeEdge(variable, compose(edgeBefore, e.modifier));
-                if (joinedEdge != null) {
-                    edges.add(joinedEdge);
-                    if (edgeBefore.source == edge.source || edgeBefore.source == variable) {
-                        stack.push(joinedEdge);
+    }
+
+    // Tries to detect cycles when a new edge is to be added.
+    // Called when a pointer propagates from variable to successor, due to an inclusion edge.
+    private List<IncludeEdge> detectCycles(Variable variable, IncludeEdge edge) {
+        // Fast check for cycles of length 1.
+        if (edge.source == variable) {
+            return List.of(edge);
+        }
+        // Fast check with lazy cycle detection:
+        // Eventually, any cycle will have a 'new' edge, where the pointer sets are equal.
+        // Therefore, we wait for this, instead of trying to immediately detect the cycle.
+        if (!equalsPointerSet(variable, edge.source)) {
+            cyclesFastCulled++;
+            return List.of();
+        }
+        // Slow check
+        Set<Variable> includerSet = new HashSet<>(List.of(variable));
+        {
+            List<Variable> queue = new ArrayList<>(List.of(variable));
+            while (!queue.isEmpty()) {
+                List<Variable> currentList = queue;
+                queue = new ArrayList<>();
+                for (Variable current : currentList) {
+                    for (Variable v : current.seeAlso) {
+                        // Culling
+                        if (includerSet.contains(v)) {
+                            continue;
+                        }
+                        // Try to find some include edge, as 'seeAlso' also indicates store and load edges.
+                        for (IncludeEdge i : v.includes) {
+                            if (i.source == current && includerSet.add(v)) {
+                                queue.add(v);
+                                break;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        if (!includerSet.contains(edge.source)) {
+            cyclesSlowCulled++;
+            return List.of();
+        }
+        cyclesDetected++;
+        // Collect all cyclic paths.
+        // Use 'set' for performance.
+        Set<IncludeEdge> set = new HashSet<>();
+        Map<Variable, List<IncludeEdge>> edges = new HashMap<>();
+        {
+            List<IncludeEdge> queue = new ArrayList<>(List.of(new IncludeEdge(edge.source, TRIVIAL_MODIFIER)));
```

Comment:
Since `set` is only used in the same scope as `queue`, you should move it there.